ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt

```diff
@@ -46,11 +55,72 @@ internal fun Offering.validatedLegacyPaywall(
         )
     }
     return PaywallValidationResult.Legacy(
-        paywallData,
+        this,
         template,
     )
 }
 
+private fun Offering.fallbackPaywall(
+    currentColorScheme: ColorScheme,
+    resourceProvider: ResourceProvider,
+    error: PaywallValidationError,
+): PaywallValidationResult.Legacy =
+    fallbackPaywall(
+        currentColorScheme = currentColorScheme,
+        resourceProvider = resourceProvider,
+        errors = nonEmptyListOf(error),
+    )
+
+private fun Offering.fallbackPaywall(
+    currentColorScheme: ColorScheme,
+    resourceProvider: ResourceProvider,
+    errors: NonEmptyList<PaywallValidationError>,
+): PaywallValidationResult.Legacy =
+    PaywallValidationResult.Legacy(
+        PaywallData.createDefault(
+            availablePackages,
+            currentColorScheme,
+            resourceProvider,
+        ),
+        PaywallData.defaultTemplate,
+        errors,
+    )
+
+@Suppress("MaxLineLength")
+internal fun PaywallComponentsData.validate(): RcResult<PaywallValidationResult.Components, NonEmptyList<PaywallValidationError>> =
+    defaultLocalization
+        // Check that the default localization is present in the localizations map.
+        .errorIfNull(PaywallValidationError.AllLocalizationsMissing(defaultLocaleIdentifier))
+        .mapError { nonEmptyListOf(it) }
+        .map { defaultLocalization ->
+            // Build a NonEmptyMap, ensuring that we always have the default localization as fallback.
+            nonEmptyMapOf(defaultLocaleIdentifier to defaultLocalization, componentsLocalizations)
+        }
+        .flatMap { localizations ->
+            // We need to turn our NonEmptyMap<LocaleId, Map> into NonEmptyMap<LocaleId, NonEmptyMap>.
+            localizations.mapValues { (locale, map) ->
+                map.toNonEmptyMapOrNull()
+                    .errorIfNull(PaywallValidationError.AllLocalizationsMissing(locale))
+                    .mapError { nonEmptyListOf(it) }
+            }.mapValuesOrAccumulate { it }
+        }.flatMap { localizations ->
+            // Use the StyleFactory to recursively create and validate all ComponentStyles.
+            val styleFactory = StyleFactory(localizations)
+            val actionHandler: suspend (PaywallAction) -> Unit = { /* TODO Move the action handler to the UI layer. */ }
+            val config = componentsConfig.base
+            zipOrAccumulate(
+                styleFactory.create(config.stack, actionHandler),
+                config.stickyFooter?.let { styleFactory.create(it, actionHandler) }.orSuccessfullyNull(),
+            ) { stack, stickyFooter ->
+                PaywallValidationResult.Components(
+                    stack = stack,
+                    stickyFooter = stickyFooter,
+                    background = config.background,
+                    locales = localizations.keys,
+                )
+            }
+        }
```

Comment:
This chain is not complex, but it's a bit hard to read, it takes some time to fully understand what's going on. I think extracting into some private functions could help

Reply:
Yea I agree. And we'll have to spend that time every time we need to read this code. Private functions is a good idea. I was also thinking of adding some early returns. I'll have a look.

Reply:
Decided to do this separately to minimize the conflicts, as I ended up making some changes to these lines in follow-up PRs. Have a look here: https://github.com/RevenueCat/purchases-android/pull/2017